pkg/trustedresources/verifier/verifier.go

```diff
@@ -0,0 +1,192 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	// keyReference is the prefix of secret reference
+	keyReference = "k8s://"
+)
+
+var (
+	// supportedSignatureAlgorithms sets a list of support signature algorithms
+	supportedSignatureAlgorithms = map[string]crypto.Hash{
+		"sha224": crypto.SHA224,
+		"sha256": crypto.SHA256,
+		"sha384": crypto.SHA384,
+		"sha512": crypto.SHA512,
+	}
+)
+
+// FromConfigMap get all verifiers from configmap, k8s is provided to fetch secret from cluster
+func FromConfigMap(ctx context.Context, k8s kubernetes.Interface) ([]signature.Verifier, error) {
+	cfg := config.FromContextOrDefaults(ctx)
+	verifiers := []signature.Verifier{}
+	for key := range cfg.TrustedResources.Keys {
+		if key == "" {
+			continue
+		}
+		v, err := fromKeyRef(ctx, key, crypto.SHA256, k8s)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get verifier from keyref: %w", err)
+		}
+		verifiers = append(verifiers, v)
+	}
+	return verifiers, nil
+}
+
+// FromPolicy get all verifiers from VerificationPolicy.
+// For each policy, loop the Authorities of the VerificationPolicy to fetch public key
+// from either inline Data or from a SecretRef.
+func FromPolicy(ctx context.Context, k8s kubernetes.Interface, policy *v1alpha1.VerificationPolicy) ([]signature.Verifier, error) {
+	verifiers := []signature.Verifier{}
+	for _, a := range policy.Spec.Authorities {
+		algorithm, err := hashAlgorithm(a.Key.HashAlgorithm)
+		if err != nil {
+			return nil, fmt.Errorf("authority %q contains an invalid hash algorithm: %w", a.Name, err)
+		}
+		if a.Key.Data == "" && a.Key.SecretRef == nil {
+			return nil, errorEmptyKey
+		}
+		if a.Key.Data != "" {
+			v, err := fromData([]byte(a.Key.Data), algorithm)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get verifier from data: %w", err)
+			}
+			verifiers = append(verifiers, v)
+		} else if a.Key.SecretRef != nil {
+			v, err := fromSecret(ctx, fmt.Sprintf("%s%s/%s", keyReference, a.Key.SecretRef.Namespace, a.Key.SecretRef.Name), algorithm, k8s)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get verifier from secret: %w", err)
+			}
+			verifiers = append(verifiers, v)
+		}
+	}
+	if len(verifiers) == 0 {
+		return verifiers, errorEmptyPublicKeys
+	}
+	return verifiers, nil
+
+}
+
+// fromKeyRef parses the given keyRef, loads the key and returns an appropriate
+// verifier using the provided hash algorithm
+func fromKeyRef(ctx context.Context, keyRef string, hashAlgorithm crypto.Hash, k8s kubernetes.Interface) (verifier signature.Verifier, err error) {
+	var raw []byte
+	if strings.HasPrefix(keyRef, keyReference) {
+		v, err := fromSecret(ctx, keyRef, hashAlgorithm, k8s)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get verifier from secret: %w", err)
+		}
+		return v, nil
+	}
+	raw, err = os.ReadFile(filepath.Clean(keyRef))
+	if err != nil {
+		return nil, fmt.Errorf("%w: %v", ErrorFailedLoadKeyFile, err)
```

Comment:
im wondering why in this case ErrorFailedLoadKeyFile isn't returned directly (it's wrapped), but other functions do return errors from errors.go directly?

Reply:
Worked through this with @Yongxuanzhang - I didn't realize that the %w syntax would wrap ErrorFailedLoadKeyFile such that the resulting error is an instance of ErrorFailedLoadKeyFile (example https://go.dev/play/p/BYiUNsGVwbA) - makes sense to me @Yongxuanzhang